File: src/engine/workers/mesher/GreedyMesher.ts
import type { ChunkSlot } from "../../alloc/SharedPool.js";
import type { BlockRegistry } from "../../../world/BlockRegistry.js";
import { calculateAO } from "./AmbientOcclusion.js";
import { shouldCullFace } from "./FaceCuller.js";
import { MAX_LIGHT, getBlockLight, getSkyLight, packLight, packVertexLight } from "./LightPacker.js";
import { getPower } from "../redstone/RedstonePacker.js";

const BLOCK_ID_MASK = 0x0fff;
const VARIANT_SHIFT = 12;

const enum Axis {
  X = 0,
  Y = 1,
  Z = 2,
}

export interface NeighborChunkSlots {
  readonly negX?: ChunkSlot;
  readonly posX?: ChunkSlot;
  readonly negZ?: ChunkSlot;
  readonly posZ?: ChunkSlot;
}

export function greedyMeshChunk(
  slot: ChunkSlot,
  dims: { sizeX: number; sizeY: number; sizeZ: number; vertexStrideFloats: number },
  blocks: BlockRegistry,
  neighborSlots: NeighborChunkSlots = {},
): boolean {
  const { sizeX, sizeY, sizeZ, vertexStrideFloats } = dims;
  const voxels = slot.voxels;
  const light = slot.light;
  const redstone = slot.redstone;
  const vertices = slot.vertices;
  const indices = slot.indices;
  let vertexCount = 0;
  let indexCount = 0;
  const maxVertices = vertices.length / vertexStrideFloats;
  const maxIndices = indices.length;

  const clampX = (x: number): number => Math.max(0, Math.min(sizeX - 1, x));
  const clampZ = (z: number): number => Math.max(0, Math.min(sizeZ - 1, z));
  const voxelAt = (source: ChunkSlot, x: number, y: number, z: number): number => (
    source.voxels[(y * sizeZ + z) * sizeX + x]
  );
  const lightAt = (source: ChunkSlot, x: number, y: number, z: number): number => (
    source.light[(y * sizeZ + z) * sizeX + x]
  );

  const get = (x: number, y: number, z: number): number => {
    if (y < 0 || y >= sizeY) return 0;
    if (x < 0) {
      return neighborSlots.negX ? voxelAt(neighborSlots.negX, sizeX - 1, y, clampZ(z)) : 0;
    }
    if (x >= sizeX) {
      return neighborSlots.posX ? voxelAt(neighborSlots.posX, 0, y, clampZ(z)) : 0;
    }
    if (z < 0) {
      return neighborSlots.negZ ? voxelAt(neighborSlots.negZ, x, y, sizeZ - 1) : 0;
    }
    if (z >= sizeZ) {
      return neighborSlots.posZ ? voxelAt(neighborSlots.posZ, x, y, 0) : 0;
    }
    return voxelAt(slot, x, y, z);
  };

  const toIndex = (x: number, y: number, z: number): number => (y * sizeZ + z) * sizeX + x;

  const getVisualVariant = (blockId: number, voxelIndex: number): number => {
    const power = getPower(redstone[voxelIndex]);
    if (blockId === 55) return power;
    if (blockId === 75 || blockId === 76 || blockId === 93 || blockId === 123) return power > 0 ? 1 : 0;
    return 0;
  };

  const encodeSurface = (blockId: number, variant: number): number => blockId | (variant << VARIANT_SHIFT);

  const faceMaskValue = (
    a: number,
    b: number,
    aOwned: boolean,
    bOwned: boolean,
    ax: number,
    ay: number,
    az: number,
    bx: number,
    by: number,
    bz: number,
  ): number => {
    if (shouldCullFace(a, b, blocks)) return 0;
    if (aOwned && a !== 0) return encodeSurface(a, getVisualVariant(a, toIndex(ax, ay, az)));
    if (bOwned && b !== 0) return -encodeSurface(b, getVisualVariant(b, toIndex(bx, by, bz)));
    return 0;
  };

  const getLight = (x: number, y: number, z: number): number => {
    if (y < 0) return packLight(0, 0);
    if (y >= sizeY) return packLight(MAX_LIGHT, 0);
    if (x < 0) {
      return neighborSlots.negX ? lightAt(neighborSlots.negX, sizeX - 1, y, clampZ(z)) : lightAt(slot, 0, y, clampZ(z));
    }
    if (x >= sizeX) {
      return neighborSlots.posX ? lightAt(neighborSlots.posX, 0, y, clampZ(z)) : lightAt(slot, sizeX - 1, y, clampZ(z));
    }
    if (z < 0) {
      return neighborSlots.negZ ? lightAt(neighborSlots.negZ, x, y, sizeZ - 1) : lightAt(slot, x, y, 0);
    }
    if (z >= sizeZ) {
      return neighborSlots.posZ ? lightAt(neighborSlots.posZ, x, y, 0) : lightAt(slot, x, y, sizeZ - 1);
    }
    return lightAt(slot, x, y, z);
  };

  const occludes = (blockId: number): number => {
    if (blockId === 0) return 0;
    const block = blocks.tryGet(blockId);
    return block && block.material.opaque ? 1 : 0;
  };

  const maxMask = Math.max(sizeY * sizeZ, sizeX * sizeZ, sizeX * sizeY);
  const mask = new Int32Array(maxMask);
  const axes = [
    { d: Axis.X, u: Axis.Y, v: Axis.Z, sizeD: sizeX, sizeU: sizeY, sizeV: sizeZ },
    { d: Axis.Y, u: Axis.X, v: Axis.Z, sizeD: sizeY, sizeU: sizeX, sizeV: sizeZ },
    { d: Axis.Z, u: Axis.X, v: Axis.Y, sizeD: sizeZ, sizeU: sizeX, sizeV: sizeY },
  ] as const;

  for (const sweep of axes) {
    const { d, u, v, sizeD, sizeU, sizeV } = sweep;
    const x = [0, 0, 0];
    const q = [0, 0, 0];
    q[d] = 1;

    for (x[d] = -1; x[d] < sizeD;) {
      let n = 0;
      for (x[v] = 0; x[v] < sizeV; x[v]++) {
        for (x[u] = 0; x[u] < sizeU; x[u]++) {
          const aOwned = x[d] >= 0;
          const bOwned = x[d] + 1 < sizeD;
          const a = get(x[0], x[1], x[2]);
          const b = get(x[0] + q[0], x[1] + q[1], x[2] + q[2]);
          mask[n++] = faceMaskValue(
            a,
            b,
            aOwned,
            bOwned,
            x[0],
            x[1],
            x[2],
            x[0] + q[0],
            x[1] + q[1],
            x[2] + q[2],
          );
        }
      }
      x[d]++;

      n = 0;
      for (let j = 0; j < sizeV; j++) {
        for (let i = 0; i < sizeU;) {
          const cell = mask[n];
          if (cell === 0) {
            i++;
            n++;
            continue;
          }

          let width = 1;
          while (i + width < sizeU && mask[n + width] === cell) width++;

          let height = 1;
          let blocked = false;
          while (j + height < sizeV) {
            for (let k = 0; k < width; k++) {
              if (mask[n + k + height * sizeU] !== cell) {
                blocked = true;
                break;
              }
            }
            if (blocked) break;
            height++;
          }

          x[u] = i;
          x[v] = j;

          const du = [0, 0, 0];
          const dv = [0, 0, 0];
          du[u] = width;
          dv[v] = height;

          const p0 = [x[0], x[1], x[2]];
          const p1 = [x[0] + du[0], x[1] + du[1], x[2] + du[2]];
          const p2 = [x[0] + du[0] + dv[0], x[1] + du[1] + dv[1], x[2] + du[2] + dv[2]];
          const p3 = [x[0] + dv[0], x[1] + dv[1], x[2] + dv[2]];

          const sign = cell > 0 ? 1 : -1;
          const surface = Math.abs(cell);
          const blockId = surface & BLOCK_ID_MASK;
          const variant = surface >>> VARIANT_SHIFT;
          const block = blocks.get(blockId);
          const normal = [0, 0, 0];
          normal[d] = sign;
          const layer = getTextureLayer(blockId, variant, block, d, sign);
          const ao = computeCornerAO(get, occludes, x, du, dv, d, u, v, sign, width, height);
          const packedLight = computeCornerLight(getLight, x, du, dv, d, u, v, sign);
          const cornerW = (i: number): number =>
            ao[i] + getSkyLight(packedLight[i]) + getBlockLight(packedLight[i]);
          const flip = Math.abs(cornerW(1) - cornerW(3)) <= Math.abs(cornerW(0) - cornerW(2));

          if (vertexCount + 4 > maxVertices || indexCount + 6 > maxIndices) {
            slot.vertexCount[0] = vertexCount;
            slot.indexCount[0] = indexCount;
            return false;
          }

          writeVertex(
            vertices,
            vertexCount * vertexStrideFloats,
            p0,
            normal,
            [0, 0],
            layer,
            packVertexLight(getSkyLight(packedLight[0]), getBlockLight(packedLight[0]), ao[0]),
          );
          writeVertex(
            vertices,
            (vertexCount + 1) * vertexStrideFloats,
            p1,
            normal,
            [width, 0],
            layer,
            packVertexLight(getSkyLight(packedLight[1]), getBlockLight(packedLight[1]), ao[1]),
          );
          writeVertex(
            vertices,
            (vertexCount + 2) * vertexStrideFloats,
            p2,
            normal,
            [width, height],
            layer,
            packVertexLight(getSkyLight(packedLight[2]), getBlockLight(packedLight[2]), ao[2]),
          );
          writeVertex(
            vertices,
            (vertexCount + 3) * vertexStrideFloats,
            p3,
            normal,
            [0, height],
            layer,
            packVertexLight(getSkyLight(packedLight[3]), getBlockLight(packedLight[3]), ao[3]),
          );

          if (flip) {
            indices[indexCount++] = vertexCount;
            indices[indexCount++] = vertexCount + 1;
            indices[indexCount++] = vertexCount + 2;
            indices[indexCount++] = vertexCount;
            indices[indexCount++] = vertexCount + 2;
            indices[indexCount++] = vertexCount + 3;
          } else {
            indices[indexCount++] = vertexCount + 1;
            indices[indexCount++] = vertexCount + 2;
            indices[indexCount++] = vertexCount + 3;
            indices[indexCount++] = vertexCount + 1;
            indices[indexCount++] = vertexCount + 3;
            indices[indexCount++] = vertexCount;
          }
          vertexCount += 4;

          for (let h = 0; h < height; h++) {
            for (let w = 0; w < width; w++) {
              mask[n + w + h * sizeU] = 0;
            }
          }

          i += width;
          n += width;
        }
      }
    }
  }

  slot.vertexCount[0] = vertexCount;
  slot.indexCount[0] = indexCount;
  return true;
}

const computeCornerAO = (
  get: (x: number, y: number, z: number) => number,
  occludes: (blockId: number) => number,
  faceOrigin: number[],
  du: number[],
  dv: number[],
  d: Axis,
  u: Axis,
  v: Axis,
  sign: number,
  width: number,
  height: number,
): [number, number, number, number] => {
  const origin = [faceOrigin[0], faceOrigin[1], faceOrigin[2]];
  if (sign > 0) origin[d] += 1;

  const sample = (uDir: number, vDir: number): number => {
    const corner = [origin[0], origin[1], origin[2]];
    if (uDir > 0) {
      corner[0] += du[0];
      corner[1] += du[1];
      corner[2] += du[2];
    }
    if (vDir > 0) {
      corner[0] += dv[0];
      corner[1] += dv[1];
      corner[2] += dv[2];
    }

    const s1 = [corner[0], corner[1], corner[2]];
    const s2 = [corner[0], corner[1], corner[2]];
    const c = [corner[0], corner[1], corner[2]];
    s1[u] += uDir;
    s2[v] += vDir;
    c[u] += uDir;
    c[v] += vDir;
    return calculateAO(
      occludes(get(s1[0] | 0, s1[1] | 0, s1[2] | 0)),
      occludes(get(s2[0] | 0, s2[1] | 0, s2[2] | 0)),
      occludes(get(c[0] | 0, c[1] | 0, c[2] | 0)),
    );
  };

  void width;
  void height;
  return [
    sample(-1, -1),
    sample(1, -1),
    sample(1, 1),
    sample(-1, 1),
  ];
};

const getTextureLayer = (
  blockId: number,
  variant: number,
  block: ReturnType<BlockRegistry["get"]>,
  d: Axis,
  sign: number,
): number => {
  if (blockId === 55) return 46 + variant;
  if (blockId === 75 || blockId === 76 || blockId === 93) return 46 + Math.min(15, variant > 0 ? 15 : 0);
  if (blockId === 123) return variant > 0 ? 63 : 62;
  return d === Axis.Y ? (sign > 0 ? block.textures.top : block.textures.bottom) : block.textures.side;
};

const computeCornerLight = (
  getLight: (x: number, y: number, z: number) => number,
  faceOrigin: number[],
  du: number[],
  dv: number[],
  d: Axis,
  u: Axis,
  v: Axis,
  sign: number,
): [number, number, number, number] => {
  const origin = [faceOrigin[0], faceOrigin[1], faceOrigin[2]];
  if (sign > 0) origin[d] += 1;

  const sample = (uDir: number, vDir: number): number => {
    const corner = [origin[0], origin[1], origin[2]];
    if (uDir > 0) {
      corner[0] += du[0];
      corner[1] += du[1];
      corner[2] += du[2];
    }
    if (vDir > 0) {
      corner[0] += dv[0];
      corner[1] += dv[1];
      corner[2] += dv[2];
    }

    const s1 = [corner[0], corner[1], corner[2]];
    const s2 = [corner[0], corner[1], corner[2]];
    const c = [corner[0], corner[1], corner[2]];
    s1[u] += uDir;
    s2[v] += vDir;
    c[u] += uDir;
    c[v] += vDir;

    const l0 = getLight(corner[0] | 0, corner[1] | 0, corner[2] | 0);
    const l1 = getLight(s1[0] | 0, s1[1] | 0, s1[2] | 0);
    const l2 = getLight(s2[0] | 0, s2[1] | 0, s2[2] | 0);
    const l3 = getLight(c[0] | 0, c[1] | 0, c[2] | 0);
    const sky = Math.round((getSkyLight(l0) + getSkyLight(l1) + getSkyLight(l2) + getSkyLight(l3)) / 4);
    const block = Math.round((getBlockLight(l0) + getBlockLight(l1) + getBlockLight(l2) + getBlockLight(l3)) / 4);
    return packLight(sky, block);
  };

  return [
    sample(-1, -1),
    sample(1, -1),
    sample(1, 1),
    sample(-1, 1),
  ];
};

const writeVertex = (
  target: Float32Array,
  offset: number,
  position: number[],
  normal: number[],
  uv: [number, number],
  layer: number,
  packedLight: number,
): void => {
  target[offset + 0] = position[0];
  target[offset + 1] = position[1];
  target[offset + 2] = position[2];
  target[offset + 3] = normal[0];
  target[offset + 4] = normal[1];
  target[offset + 5] = normal[2];
  target[offset + 6] = uv[0];
  target[offset + 7] = uv[1];
  target[offset + 8] = layer;
  target[offset + 9] = packedLight;
};

File: src/engine/workers/mesher/GreedyMesher.cpp
#include "GreedyMesher.hpp"
#include "AmbientOcclusion.hpp"
#include "LightSampling.hpp"
#include "world/BlockIds.hpp"
#include <algorithm>
#include <cstring>
#include <vector>

// @deprecated Legacy voxel-world code retained during the render-only migration to triangle meshes.
namespace voxel {
namespace mesher {

namespace {

constexpr uint32_t kPackedVertexFloats = 10u;

} // namespace

// ======================================================================
// Vertex writer
// ======================================================================

static inline void writeVtx(float* data, uint32_t offset,
                            float x, float y, float z,
                            float nx, float ny, float nz,
                            float u, float v,
                            float texLayer, float lightData) {
  float* p = data + offset;
  p[0]=x; p[1]=y; p[2]=z;       // a_pos
  p[3]=nx; p[4]=ny; p[5]=nz;    // a_normal
  p[6]=u; p[7]=v;               // a_uv
  p[8]=texLayer;                 // a_texLayer
  p[9]=lightData;                // a_lightData
}

namespace {
struct GreedyMeshScratch {
  std::vector<uint16_t> mask;
  std::vector<uint32_t> transparentIndices;
};
thread_local GreedyMeshScratch g_scratch;

auto isOverlayRenderableBlock(const BlockDefinition& def) -> bool {
  switch (def.id) {
    case BlockId::GRASS:
    case BlockId::DIRT:
    case BlockId::STONE:
    case BlockId::SAND:
    case BlockId::BEDROCK:
    case BlockId::MOSSY_STONE:
      return false;
    default:
      return true;
  }
}
}

// ======================================================================
// Public API
// ======================================================================

auto estimateMeshCapacity(
    const uint8_t* voxels,
    const BlockRegistry& blocks,
    const MesherConfig& cfg,
    const NeighborVoxelViews& neighbors) -> MeshCapacityHint
{
  MeshCapacityHint hint{};
  if (!voxels) return hint;
  if (cfg.sizeX <= 0 || cfg.sizeY <= 0 || cfg.sizeZ <= 0) return hint;

  static constexpr int32_t kNeighborOffsets[6][3] = {
    { 1, 0, 0}, {-1, 0, 0},
    { 0, 1, 0}, { 0,-1, 0},
    { 0, 0, 1}, { 0, 0,-1},
  };

  for (int32_t y = 0; y < cfg.sizeY; ++y) {
    for (int32_t z = 0; z < cfg.sizeZ; ++z) {
      for (int32_t x = 0; x < cfg.sizeX; ++x) {
        uint8_t bid = getBlock(voxels, x, y, z, cfg, neighbors);
        if (bid == 0) continue;

        const auto* def = blocks.tryGet(bid);
        if (!def) continue;

        for (const auto& off : kNeighborOffsets) {
          const int32_t nx = x + off[0];
          const int32_t ny = y + off[1];
          const int32_t nz = z + off[2];
          if (isOpaqueBlock(getBlock(voxels, nx, ny, nz, cfg, neighbors), blocks)) {
            continue;
          }
          ++hint.quadCount;
        }
      }
    }
  }

  hint.vertexCount = hint.quadCount * 4u;
  hint.indexCount = hint.quadCount * 6u;
  return hint;
}

// @see notes/mesher-capacity-accounting.md
template <typename ShouldEmitFn>
bool meshVolumeImpl(
    const uint8_t* voxels,
    const uint8_t* light,
    const BlockRegistry& blocks,
    const MesherConfig& cfg,
    uint32_t vertexBase,
    uint32_t opaqueIndexBase,
    float* vertexOut,
    uint32_t* indexOut,
    uint32_t& vertexCountOut,
    uint32_t& indexCountOut,
    bool* hasTransparentOut,
    bool* hasOpaqueOut,
    uint32_t* opaqueIndexCountOut,
    uint32_t* transparentIndexCountOut,
    const NeighborVoxelViews& neighbors,
    ShouldEmitFn&& shouldEmitBlock)
{
  vertexCountOut = vertexBase;
  indexCountOut = opaqueIndexBase;
  if (hasTransparentOut) *hasTransparentOut = false;
  if (hasOpaqueOut) *hasOpaqueOut = opaqueIndexBase > 0u;
  if (opaqueIndexCountOut) *opaqueIndexCountOut = opaqueIndexBase;
  if (transparentIndexCountOut) *transparentIndexCountOut = 0;

  if (!voxels || !light || !vertexOut || !indexOut) return false;
  if (cfg.sizeX <= 0 || cfg.sizeY <= 0 || cfg.sizeZ <= 0) return false;
  if (cfg.maxVertices < 0 || cfg.maxIndices < 0) return false;
  if (cfg.strideFloats < static_cast<int32_t>(kPackedVertexFloats)) return false;

  const int32_t SX = cfg.sizeX;
  const int32_t SY = cfg.sizeY;
  const int32_t SZ = cfg.sizeZ;
  const int32_t S  = cfg.strideFloats; // stride in floats (should be 10)
  const uint32_t maxVertices = static_cast<uint32_t>(cfg.maxVertices);
  const uint32_t maxI  = static_cast<uint32_t>(cfg.maxIndices);
  const uint32_t strideFloats = static_cast<uint32_t>(S);

  if (vertexBase > maxVertices || opaqueIndexBase > maxI) return false;

  uint32_t vo = vertexBase * strideFloats; // float offset into vertexOut
  uint32_t opaqueIo = opaqueIndexBase;      // opaque index count

  // Scratch mask sized to the largest face for this chunk configuration.
  const auto neededScratch =
    static_cast<size_t>(std::max({SX * SY, SX * SZ, SY * SZ}));
  if (g_scratch.mask.size() < neededScratch) {
    g_scratch.mask.resize(neededScratch);
  }
  g_scratch.transparentIndices.clear();
  uint16_t* mask = g_scratch.mask.data();

  for (int32_t di = 0; di < 6; ++di) {
    const auto& info = kDirs[di];
    int32_t axis  = info.axis;
    int32_t uAxis = info.uAxis;
    int32_t vAxis = info.vAxis;
    int32_t sign  = info.sign;

    float nx = static_cast<float>(info.normal[0]);
    float ny = static_cast<float>(info.normal[1]);
    float nz = static_cast<float>(info.normal[2]);

    int32_t sizes[3] = {SX, SY, SZ};
    int32_t sliceCnt = sizes[axis];
    int32_t uSz      = sizes[uAxis];
    int32_t vSz      = sizes[vAxis];

    size_t needed = static_cast<size_t>(uSz) * vSz;

    for (int32_t sl = 0; sl < sliceCnt; ++sl) {
      // ---- Build mask ----
      std::memset(mask, 0, needed * sizeof(uint16_t));

      for (int32_t v = 0; v < vSz; ++v) {
        for (int32_t u = 0; u < uSz; ++u) {
          int32_t co[3];
          co[axis]  = sl;
          co[uAxis] = u;
          co[vAxis] = v;
          int32_t bx = co[0], by = co[1], bz = co[2];

          uint8_t bid = getBlock(voxels, bx, by, bz, cfg, neighbors);
          if (bid == 0) continue;

          int32_t nn[3] = {bx + info.normal[0], by + info.normal[1], bz + info.normal[2]};
          uint8_t nid = getBlock(voxels, nn[0], nn[1], nn[2], cfg, neighbors);
          if (nid != 0) {
            const auto* nd = blocks.tryGet(nid);
            if (nd && nd->material.opaque) continue;
          }
          uint8_t aoKey = computeFaceAOPacked(voxels, di, sl, u, v, info, cfg, blocks, neighbors);
          mask[v * uSz + u] = static_cast<uint16_t>(bid) | (static_cast<uint16_t>(aoKey) << 8);
        }
      }

      // ---- Greedy rectangle packing ----
      for (int32_t v = 0; v < vSz; ++v) {
        for (int32_t u = 0; u < uSz; ++u) {
          uint16_t maskVal = mask[v * uSz + u];
          if (maskVal == 0) continue;
          uint8_t bid = static_cast<uint8_t>(maskVal & 0xFF);

          const auto* def = blocks.tryGet(bid);
          if (!def || !shouldEmitBlock(*def)) { mask[v * uSz + u] = 0; continue; }
          if (hasTransparentOut && !def->material.opaque) *hasTransparentOut = true;
          if (hasOpaqueOut && def->material.opaque) *hasOpaqueOut = true;

          // Texture layer for this face direction
          uint16_t tlayer;
          switch (di) {
            case 0: case 1: tlayer = def->textures.side;   break;
            case 2:         tlayer = def->textures.top;    break;
            case 3:         tlayer = def->textures.bottom; break;
            case 4: case 5: tlayer = def->textures.side;   break;
            default:        tlayer = 0; break;
          }

          // Grow rectangle
          int32_t rw = 1, rh = 1;

          while (u + rw < uSz) {
            bool ok = true;
            for (int32_t vv = 0; vv < rh; ++vv)
              if (mask[(v+vv)*uSz + u+rw] != maskVal) { ok = false; break; }
            if (!ok) break;
            ++rw;
          }
          while (v + rh < vSz) {
            bool ok = true;
            for (int32_t uu = 0; uu < rw; ++uu)
              if (mask[(v+rh)*uSz + u+uu] != maskVal) { ok = false; break; }
            if (!ok) break;
            ++rh;
          }

          // Clear mask
          for (int32_t vv = 0; vv < rh; ++vv)
            for (int32_t uu = 0; uu < rw; ++uu)
              mask[(v+vv)*uSz + u+uu] = 0;

          // ----- Emit quad -----
          float fc = static_cast<float>(sl) + (sign > 0 ? 1.0f : 0.0f);

          // 4 corners
          float bl[3]={}, br[3]={}, tl[3]={}, tr[3]={};
          bl[axis] = br[axis] = tl[axis] = tr[axis] = fc;
          bl[uAxis] = static_cast<float>(u);       br[uAxis] = static_cast<float>(u+rw);
          tl[uAxis] = static_cast<float>(u);       tr[uAxis] = static_cast<float>(u+rw);
          bl[vAxis] = static_cast<float>(v);       br[vAxis] = static_cast<float>(v);
          tl[vAxis] = static_cast<float>(v+rh);    tr[vAxis] = static_cast<float>(v+rh);

          float uv[4][2] = {
            {0,0}, {static_cast<float>(rw),0},
            {0,static_cast<float>(rh)}, {static_cast<float>(rw),static_cast<float>(rh)},
          };

          // Grid corner integers
          int32_t c[4][3] = {
            {static_cast<int32_t>(bl[0]), static_cast<int32_t>(bl[1]), static_cast<int32_t>(bl[2])},
            {static_cast<int32_t>(br[0]), static_cast<int32_t>(br[1]), static_cast<int32_t>(br[2])},
            {static_cast<int32_t>(tl[0]), static_cast<int32_t>(tl[1]), static_cast<int32_t>(tl[2])},
            {static_cast<int32_t>(tr[0]), static_cast<int32_t>(tr[1]), static_cast<int32_t>(tr[2])},
          };

          // Extract pre-computed AO from the mask key — all faces in this
          // merged rect are guaranteed to share the same AO pattern.
          uint8_t aoKey = static_cast<uint8_t>((maskVal >> 8) & 0xFF);
          int32_t ao[4], sky[4], block[4];
          for (int32_t ci = 0; ci < 4; ++ci) {
            ao[ci] = (aoKey >> (ci * 2)) & 0x3;
            AvgLight al = cornerLight(light, axis, sign, uAxis, vAxis, c[ci], cfg);
            sky[ci] = al.sky;
            block[ci] = al.block;
          }

          float ld[4];
          for (int32_t ci = 0; ci < 4; ++ci)
            ld[ci] = packLight(sky[ci], block[ci], ao[ci]);

          bool flip = (ao[0] + ao[2]) > (ao[1] + ao[3]);

          // Capacity
          const uint32_t nextVertexCount = (vo / strideFloats) + 4u;
          const bool faceOpaque = def->material.opaque;
          const uint32_t transparentIo = static_cast<uint32_t>(g_scratch.transparentIndices.size());
          const uint32_t nextIndexCount = opaqueIo + transparentIo + 6u;
          if (nextVertexCount > maxVertices || nextIndexCount > maxI) {
            vertexCountOut = vo / strideFloats;
            indexCountOut = opaqueIo + transparentIo;
            if (opaqueIndexCountOut) *opaqueIndexCountOut = opaqueIo;
            if (transparentIndexCountOut) *transparentIndexCountOut = transparentIo;
            return false;
          }

          float tlF = static_cast<float>(tlayer);
          writeVtx(vertexOut, vo, bl[0],bl[1],bl[2], nx,ny,nz, uv[0][0],uv[0][1], tlF, ld[0]); vo += S;
          writeVtx(vertexOut, vo, br[0],br[1],br[2], nx,ny,nz, uv[1][0],uv[1][1], tlF, ld[1]); vo += S;
          writeVtx(vertexOut, vo, tl[0],tl[1],tl[2], nx,ny,nz, uv[2][0],uv[2][1], tlF, ld[2]); vo += S;
          writeVtx(vertexOut, vo, tr[0],tr[1],tr[2], nx,ny,nz, uv[3][0],uv[3][1], tlF, ld[3]); vo += S;

          uint32_t base = (vo / strideFloats) - 4;
          auto emitIndex = [&](uint32_t idx) {
            if (faceOpaque) {
              indexOut[opaqueIo++] = idx;
            } else {
              g_scratch.transparentIndices.push_back(idx);
            }
          };

          // Faces X+ (di=0), Y+ (di=2), and Z- (di=5) produce a computed
          // normal opposite to the face normal with the default winding.
          // Reverse the winding for those three directions so that the
          // geometric normal matches the per-vertex normal, giving correct
          // diffuse lighting.
          bool reverseWind = (di == 0 || di == 2 || di == 5);

          if (!flip) {
            if (!reverseWind) {
              emitIndex(base);   emitIndex(base+1); emitIndex(base+2);
              emitIndex(base+1); emitIndex(base+3); emitIndex(base+2);
            } else {
              emitIndex(base);   emitIndex(base+2); emitIndex(base+3);
              emitIndex(base+1); emitIndex(base);   emitIndex(base+3);
            }
          } else {
            if (!reverseWind) {
              emitIndex(base);   emitIndex(base+1); emitIndex(base+3);
              emitIndex(base);   emitIndex(base+3); emitIndex(base+2);
            } else {
              emitIndex(base);   emitIndex(base+2); emitIndex(base+1);
              emitIndex(base+2); emitIndex(base+3); emitIndex(base+1);
            }
          }
        }
      }
    }
  }

  const uint32_t transparentIo = static_cast<uint32_t>(g_scratch.transparentIndices.size());
  if (transparentIo > 0u) {
    std::memcpy(indexOut + opaqueIo, g_scratch.transparentIndices.data(),
                static_cast<size_t>(transparentIo) * sizeof(uint32_t));
  }
  vertexCountOut = vo / strideFloats;
  indexCountOut  = opaqueIo + transparentIo;
  if (opaqueIndexCountOut) *opaqueIndexCountOut = opaqueIo;
  if (transparentIndexCountOut) *transparentIndexCountOut = transparentIo;
  return true;
}

bool greedyMesh(
    const uint8_t* voxels,
    const uint8_t* light,
    const BlockRegistry& blocks,
    const MesherConfig& cfg,
    float* vertexOut,
    uint32_t* indexOut,
    uint32_t& vertexCountOut,
    uint32_t& indexCountOut,
    bool* hasTransparentOut,
    bool* hasOpaqueOut,
    uint32_t* opaqueIndexCountOut,
    uint32_t* transparentIndexCountOut,
    const NeighborVoxelViews& neighbors)
{
  return meshVolumeImpl(
      voxels, light, blocks, cfg,
      0u, 0u,
      vertexOut, indexOut,
      vertexCountOut, indexCountOut,
      hasTransparentOut, hasOpaqueOut,
      opaqueIndexCountOut, transparentIndexCountOut,
      neighbors,
      [](const BlockDefinition&) { return true; });
}

bool overlayGreedyMesh(
    const uint8_t* voxels,
    const uint8_t* light,
    const BlockRegistry& blocks,
    const MesherConfig& cfg,
    uint32_t vertexBase,
    uint32_t opaqueIndexBase,
    float* vertexOut,
    uint32_t* indexOut,
    uint32_t& vertexCountOut,
    uint32_t& indexCountOut,
    bool* hasTransparentOut,
    bool* hasOpaqueOut,
    uint32_t* opaqueIndexCountOut,
    uint32_t* transparentIndexCountOut,
    const NeighborVoxelViews& neighbors)
{
  return meshVolumeImpl(
      voxels, light, blocks, cfg,
      vertexBase, opaqueIndexBase,
      vertexOut, indexOut,
      vertexCountOut, indexCountOut,
      hasTransparentOut, hasOpaqueOut,
      opaqueIndexCountOut, transparentIndexCountOut,
      neighbors,
      [](const BlockDefinition& def) { return isOverlayRenderableBlock(def); });
}

} // namespace mesher
} // namespace voxel
